src/properties/ecc.c

```diff
@@ -0,0 +1,344 @@
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_vector_int_t degree;
+    igraph_lazy_adjlist_t al;
+    igraph_vector_bool_t b;
+    igraph_eit_t eit;
+    const igraph_real_t c = offset ? 1.0 : 0.0;
+
+    IGRAPH_CHECK(igraph_lazy_adjlist_init(graph, &al, IGRAPH_ALL, IGRAPH_NO_LOOPS, IGRAPH_NO_MULTIPLE));
+    IGRAPH_FINALLY(igraph_lazy_adjlist_destroy, &al);
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&degree, no_of_nodes);
+    IGRAPH_CHECK(igraph_degree(graph, &degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS));
+
+    IGRAPH_CHECK(igraph_eit_create(graph, eids, &eit));
+    IGRAPH_FINALLY(igraph_eit_destroy, &eit);
+
+    IGRAPH_CHECK(igraph_vector_resize(res, IGRAPH_EIT_SIZE(eit)));
+
+    IGRAPH_VECTOR_BOOL_INIT_FINALLY(&b, no_of_edges);
+
+    for (igraph_integer_t i=0;
+         ! IGRAPH_EIT_END(eit);
+         IGRAPH_EIT_NEXT(eit), i++) {
+
+        igraph_integer_t edge = IGRAPH_EIT_GET(eit);
+        igraph_integer_t v1 = IGRAPH_FROM(graph, edge), v2 = IGRAPH_TO(graph, edge);
+
+        igraph_real_t z; /* number of triangles the edge participates in */
+        igraph_real_t s; /* max number of triangles the edge could be part of */
+
+        IGRAPH_ALLOW_INTERRUPTION();
+
+        if (v1 == v2) {
+            /* A self-loop isn't, and cannot be part of any triangles. */
+            z = 0.0;
+            s = 0.0;
+        } else {
+            igraph_vector_int_t *a1 = igraph_lazy_adjlist_get(&al, v1);
+            if (! VECTOR(b)[edge]) {
+                IGRAPH_CHECK_OOM(a1, "Not enough memory for edge clustering coefficient.");
+                igraph_vector_int_sort(a1);
+                VECTOR(b)[edge] = true;
+            }
+
+            igraph_vector_int_t *a2 = igraph_lazy_adjlist_get(&al, v2);
+            if (! VECTOR(b)[edge]) {
+                IGRAPH_CHECK_OOM(a2, "Not enough memory for edge clustering coefficient.");
+                igraph_vector_int_sort(a1);
+                VECTOR(b)[edge] = true;
+            }
+
+            igraph_integer_t d1 = VECTOR(degree)[v1], d2 = VECTOR(degree)[v2];
+
+            z = vector_int_intersection_size_sorted(a1, a2);
+            s = (d1 < d2 ? d1 : d2) - 1.0;
+        }
+
+        VECTOR(*res)[i] = z + c;
+        if (normalize) VECTOR(*res)[i] /= s;
+    }
+
+    igraph_vector_bool_destroy(&b);
+    igraph_eit_destroy(&eit);
+    igraph_vector_int_destroy(&degree);
+    igraph_lazy_adjlist_destroy(&al);
+    IGRAPH_FINALLY_CLEAN(4);
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t igraph_i_ecc4(
+        const igraph_t *graph, igraph_vector_t *res,
+        const igraph_es_t eids, igraph_bool_t offset, igraph_bool_t normalize) {
+
+    igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    igraph_vector_int_t degree;
+    igraph_adjlist_t al;
+    igraph_eit_t eit;
+    igraph_real_t c = offset ? 1.0 : 0.0;
+
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &al, IGRAPH_ALL, IGRAPH_NO_LOOPS, IGRAPH_NO_MULTIPLE));
```

Comment:
The adjacency list construction here is the same as in `igraph_i_ecc3()`, right? Wouldn't it be better to move it out to a separate static function so we don't need to repeat it twice? (I mean both the construction and the sorting of its elements).

Reply:
First I'll do a benchmark to see if there is even a point in keeping separate implementations for doing the computation for _all_ edges vs _some_ edges.

Reply:
I didn't look carefully and I misunderstood. I thought you were asking to share code between the two versions of `igraph_ecc3`, but you are asking to share code between `ecc3_1` and `ecc4`.

I consider the implementations of the `k=3` and `k=4` cases to be entirely separate and mostly unrelated. I worked on them separately and will continue to work on them separately. Unless we embark on implementing this for arbitrarily large `k` (which is problematic, so unlikely) I would rather not share code between them.